test/Unread-test.ts

```diff
@@ -19,100 +19,319 @@ import {
+            expect(doesRoomHaveUnreadMessages(room)).toBe(true);
+        });
+
+        it('returns false for a room when the latest event was sent by the current user', () => {
+            event = mkEvent({
+                event: true,
+                type: "m.room.message",
+                user: client.getUserId(),
+                room: roomId,
+                content: {},
+            });
+            // Only for timeline events.
+            room.addLiveEvents([event]);
+
+            expect(doesRoomHaveUnreadMessages(room)).toBe(false);
+        });
+
+        it('returns false for a room when the read receipt is at the latest event', () => {
+            const receipt = new MatrixEvent({
+                type: "m.receipt",
+                room_id: "!foo:bar",
+                content: {
+                    [event.getId()]: {
+                        [ReceiptType.Read]: {
+                            [client.getUserId()]: { ts: 1 },
+                        },
+                    },
+                },
+            });
+            room.addReceipt(receipt);
+
+            expect(doesRoomHaveUnreadMessages(room)).toBe(false);
+        });
+
+        it('returns true for a room when the read receipt is earlier than the latest event', () => {
+            const receipt = new MatrixEvent({
+                type: "m.receipt",
+                room_id: "!foo:bar",
+                content: {
+                    [event.getId()]: {
+                        [ReceiptType.Read]: {
+                            [client.getUserId()]: { ts: 1 },
+                        },
+                    },
+                },
+            });
+            room.addReceipt(receipt);
+
+            const event2 = mkEvent({
+                event: true,
+                type: "m.room.message",
+                user: aliceId,
+                room: roomId,
+                content: {},
+            });
+            // Only for timeline events.
+            room.addLiveEvents([event2]);
+
+            expect(doesRoomHaveUnreadMessages(room)).toBe(true);
+        });
+
+        it('returns true for a room with an unread message in a thread', () => {
+            // Mark the main timeline as read.
+            const receipt = new MatrixEvent({
+                type: "m.receipt",
+                room_id: "!foo:bar",
+                content: {
+                    [event.getId()]: {
+                        [ReceiptType.Read]: {
+                            [client.getUserId()]: { ts: 1 },
+                        },
+                    },
+                },
+            });
+            room.addReceipt(receipt);
+
+            // Create a thread as a different user.
+            mkThread({ room, client, authorId: client.getUserId(), participantUserIds: [aliceId] });
+
+            expect(doesRoomHaveUnreadMessages(room)).toBe(true);
+        });
+
+        it('returns false for a room when the latest thread event was sent by the current user', () => {
```

Comment:
This test is failing -- it seems like the thread events aren't being properly registered? (The thread timeline is empty.) Any thoughts of what could be going wrong?

Reply:
I've just played around with that a bit and saw the same behaviour as you, which is surprising.
@justjanne do you have any idea on why this might be happening?

Reply:
You don’t set supportsExperimentalThreads on the client. Without that, you’re not able to add events to threads, as eventShouldLiveIn always returns room

Reply:
Additionally, you'll want to modify the `test-utils/threads.ts` file in line 129

```diff
diff --git a/test/Unread-test.ts b/test/Unread-test.ts
index 8de7fa4f01..030ef91805 100644
--- a/test/Unread-test.ts
+++ b/test/Unread-test.ts
@@ -140,6 +140,10 @@ describe("Unread", () => {
         stubClient();
         const client = MatrixClientPeg.get();
 
+        beforeAll(() => {
+            client.supportsExperimentalThreads = () => true;
+        })
+
         beforeEach(() => {
             // Create a room and initial event in it.
             room = new Room(roomId, client, client.getUserId());
@@ -255,7 +259,7 @@ describe("Unread", () => {
             room.addReceipt(receipt);
 
             // Create a thread as the current user..
-            mkThread({ room, client, authorId: client.getUserId(), participantUserIds: [client.getUserId()] });
+            const { thread } = mkThread({ room, client, authorId: client.getUserId(), participantUserIds: [client.getUserId()] });
 
             expect(doesRoomHaveUnreadMessages(room)).toBe(false);
         });
diff --git a/test/test-utils/threads.ts b/test/test-utils/threads.ts
index 3b07c45051..db4a51cc71 100644
--- a/test/test-utils/threads.ts
+++ b/test/test-utils/threads.ts
@@ -126,7 +126,7 @@ export const mkThread = ({
     const thread = room.createThread(rootEvent.getId(), rootEvent, events, true);
     // So that we do not have to mock the thread loading
     thread.initialEventsFetched = true;
-    thread.addEvents(events, true);
+    thread.addEvents(events, false);
 
     return { thread, rootEvent, events };
 };
```

Reply:
That worked, thank you!